ggml.h

```diff
@@ -2367,6 +2369,90 @@ extern "C" {
     GGML_API int ggml_cpu_has_vsx        (void);
     GGML_API int ggml_cpu_has_matmul_int8(void);
 
+    /**
+     * Google Brain 16-bit floating point number.
+     *
+     *       ┌sign
+     *       │
+     *       │   ┌exponent
+     *       │   │
+     *       │   │      ┌mantissa
+     *       │   │      │
+     *       │┌──┴───┐┌─┴───┐
+     *     0b0000000000000000 brain16
+     *
+     * Since bf16 has the same number of exponent bits as a 32bit float,
+     * encoding and decoding numbers becomes relatively straightforward.
+     *
+     *       ┌sign
+     *       │
+     *       │   ┌exponent
+     *       │   │
+     *       │   │      ┌mantissa
+     *       │   │      │
+     *       │┌──┴───┐┌─┴───────────────────┐
+     *     0b00000000000000000000000000000000 IEEE binary32
+     *
+     * For comparison, the standard fp16 format has fewer exponent bits.
+     *
+     *       ┌sign
+     *       │
+     *       │  ┌exponent
+     *       │  │
+     *       │  │    ┌mantissa
+     *       │  │    │
+     *       │┌─┴─┐┌─┴──────┐
+     *     0b0000000000000000 IEEE binary16
+     *
+     * So be warned that converting between them, destroys several bits.
+     *
+     * @see IEEE 754-2008
+     */
+    typedef struct {
+        uint16_t x;
+    } ggml_bf16_t;
```

Comment:
Maybe for consistency change to:

```suggestion
    typedef uint16_t ggml_bf16_t;
```

Though, I can also see advantage of the proposed strong-typed struct. So maybe see if we can apply the same for `ggml_fp16_t`.

Reply:
I would be happy to convert `ggml_fp16_t` to a non-integral type, but I can't do that in this PR. The interplay between `ggml_fp16_t` and `ggml_half` would require a nontrivial amount of thought to resolve. In general it looks error-prone. If you ask me to fix this, I will send a follow-up PR doing just that.

Reply:
No worries - I don't think it is high priority, since we don't expect more integral types to be introduced and BF16 is already a non-integral. So we would like to change `ggml_fp16_t` mostly for consistency, which is nice but not too important